src/newgrf_badge_config.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file newgrf_badge_config.cpp Functionality for NewGRF badge configuration. */
+
+#include "stdafx.h"
+#include "core/string_consumer.hpp"
+#include "ini_type.h"
+#include "newgrf.h"
+#include "newgrf_badge.h"
+#include "newgrf_badge_config.h"
+#include "newgrf_badge_type.h"
+
+#include "table/strings.h"
+
+#include "safeguards.h"
+
+/** Global state for badge class configuration. */
+class BadgeClassConfig {
+public:
+	std::array<std::vector<BadgeClassConfigItem>, GrfSpecFeature::GSF_END> features = {};
+
+	static constexpr GrfSpecFeatures CONFIGURABLE_FEATURES = {
+		GSF_TRAINS,     GSF_ROADVEHICLES, GSF_SHIPS,   GSF_AIRCRAFT,  GSF_STATIONS,     GSF_BRIDGES,   GSF_HOUSES,    GSF_INDUSTRYTILES,
+		GSF_INDUSTRIES, GSF_AIRPORTS,     GSF_OBJECTS, GSF_RAILTYPES, GSF_AIRPORTTILES, GSF_ROADTYPES, GSF_TRAMTYPES, GSF_ROADSTOPS,
+	};
+
+	static inline const std::array<std::string_view, GrfSpecFeature::GSF_END> sections = {
+		"badges_trains", // GSF_TRAINS
+		"badges_roadvehicles", // GSF_ROADVEHICLES
+		"badges_ships", // GSF_SHIPS
+		"badges_aircraft", // GSF_AIRCRAFT
+		"badges_stations", // GSF_STATIONS
+		{}, // GSF_CANALS
+		"badges_bridges", // GSF_BRIDGES
+		"badges_houses", // GSF_HOUSES
+		{}, // GSF_GLOBALVAR
+		"badges_industrytiles", // GSF_INDUSTRYTILES
+		"badges_industries", // GSF_INDUSTRIES
+		{}, // GSF_CARGOES
+		{}, // GSF_SOUNDFX
+		"badge_columns_airports", // GSF_AIRPORTS
```

Comment:
This one does not match the rest:
```suggestion
		"badges_airports", // GSF_AIRPORTS
```